pkg/vpcmodel/output.go

```diff
@@ -103,33 +109,89 @@ type SingleAnalysisOutput struct {
 	case ARCHDRAWIO:
 		formatter = &ArchDrawioOutputFormatter{}
-	case Debug:
-		formatter = &DebugOutputFormatter{}
 	default:
-		return nil, errors.New("unsupported output format")
+		return "", errors.New("unsupported output format")
 	}
-
 	return formatter.WriteOutput(o.config1, o.config2, o.nodesConn, o.subnetsConn, o.cfgsDiff, outFile, o.outputGrouping, o.useCase)
 }
 
-type OutputFormatter interface {
-	WriteOutput(c1, c2 *VPCConfig, conn *VPCConnectivity, subnetsConn *VPCsubnetConnectivity, subnetsDiff *diffBetweenCfgs,
+// SingleVpcOutputFormatter is an interface for a formatter that can handle only one vpc
+// this interface is implemented by textOutputFormatter, jsonOutputFormatter, mdOutputFormatter
+type SingleVpcOutputFormatter interface {
+	WriteOutput(c1, c2 *VPCConfig, conn *VPCConnectivity,
+		subnetsConn *VPCsubnetConnectivity, subnetsDiff *diffBetweenCfgs,
 		outFile string, grouping bool, uc OutputUseCase) (*SingleAnalysisOutput, error)
 }
 
+// OutputFormatter is an interface for formatter that handle multi vpcs.
+// implemented by serialOutputFormatter and drawioOutputFormatter
+type OutputFormatter interface {
+	WriteOutput(c1, c2 map[string]*VPCConfig, conn map[string]*VPCConnectivity,
+		subnetsConn map[string]*VPCsubnetConnectivity, subnetsDiff *diffBetweenCfgs,
+		outFile string, grouping bool, uc OutputUseCase) (string, error)
+}
+
+// serialOutputFormatter is the formatter for json, md, txt formats.
+// serialOutputFormatter implements the interface OutputFormatter.
+// its hold an instance of a formater that implement
+// its main flow of WriteOutput() of serialOutputFormatter is:
+// 1. for each vpc, use a SingleVpcOutputFormatter to create a VPCsubnetConnectivity,
+// 2. aggregate the VPCsubnetConnectivity to one output
+type serialOutputFormatter struct {
+	outFormat OutFormat
+}
+
+func (of *serialOutputFormatter) createSingleVpcFormatter() SingleVpcOutputFormatter {
+	switch of.outFormat {
+	case JSON:
+		return &JSONoutputFormatter{}
+	case Text:
+		return &TextOutputFormatter{}
+	case MD:
+		return &MDoutputFormatter{}
+	case Debug:
+		return &DebugOutputFormatter{}
+	}
+	return nil
+}
+
+func (of *serialOutputFormatter) WriteOutput(c1, c2 map[string]*VPCConfig, conns map[string]*VPCConnectivity,
+	subnetsConns map[string]*VPCsubnetConnectivity, subnetsDiff *diffBetweenCfgs,
+	outFile string, grouping bool, uc OutputUseCase) (string, error) {
+	diffAnalysis := uc == EndpointsDiff || uc == SubnetsDiff
+	if !diffAnalysis {
+		outputPerVPC := make([]*SingleAnalysisOutput, len(c1))
+		i := 0
+		for name := range c1 {
+			vpcAnalysisOutput, err2 :=
+				of.createSingleVpcFormatter().WriteOutput(c1[name], nil, conns[name], subnetsConns[name], subnetsDiff, "", grouping, uc)
+			if err2 != nil {
+				return "", err2
+			}
+			outputPerVPC[i] = vpcAnalysisOutput
+			i++
+		}
+		return of.AggregateVPCsOutput(outputPerVPC, uc, outFile)
+	}
+	// its a diff mode, we have only one vpc on each map:
```

Comment:
Is it really require to treat separately the diff case?

Reply:
yes, in diff case we do not call AggregateVPCsOutput() 
and we call WriteDiffOutput()